File: mvp_app/MyProject/libtsparser/src/main/java/com/tosmart/xiexuming/libtsparser/section/ProgramAssociationSectionManager.java
package com.tosmart.xiexuming.libtsparser.section;


import com.tosmart.xiexuming.libtsparser.tspacket.TsPacket;

import java.util.Observable;
import java.util.Observer;


/**
 * @author xiexuming
 */
public class ProgramAssociationSectionManager extends SectionManager implements Observer {
    private static final int PROGRAM_ASSOCIATION_SPECIAL_HEAD_LENGTH = 5;
    private static final int CRC_CODE_LENGTH = 4;
    private static final int SPECIFIED_PAT_TABLE_ID = 0;
    private static final int PAT_PID = 0;

    private ProgramAssociationSection mProgramAssociationSection;

    @Override
    public void composeSpecifiedSection(Section commonSection, Integer pid) {
        if (commonSection.getCurrentLength() == commonSection.getSectionLength()) {
            byte[] commonSectionContent = commonSection.getData();
            int[] programAssociationFilledHead = new int[8];
            programAssociationFilledHead[0] = commonSection.getTableId();
            programAssociationFilledHead[1] = commonSection.getReserved();
            programAssociationFilledHead[2] = commonSection.getSectionLength();
            programAssociationFilledHead[3] = commonSectionContent[0] << 8 | commonSectionContent[1] & 0xFF;
            programAssociationFilledHead[4] = commonSectionContent[2] >> 1 & 0x1F;
            programAssociationFilledHead[5] = commonSectionContent[2] & 0x01;
            programAssociationFilledHead[6] = commonSectionContent[3];
            programAssociationFilledHead[7] = commonSectionContent[4];
            ProgramAssociationSection programAssociationSection = new ProgramAssociationSection();
            programAssociationSection.setHeadDate(programAssociationFilledHead);

            int copyProgramLength = commonSectionContent.length - PROGRAM_ASSOCIATION_SPECIAL_HEAD_LENGTH - CRC_CODE_LENGTH;
            byte[] programContent = new byte[copyProgramLength];
            System.arraycopy(commonSectionContent, PROGRAM_ASSOCIATION_SPECIAL_HEAD_LENGTH, programContent, 0, copyProgramLength);
            programAssociationSection.setData(programContent);

            int crcStartLocation = commonSectionContent.length - CRC_CODE_LENGTH;
            byte[] crcCode = new byte[4];
            crcCode[0] = commonSectionContent[crcStartLocation];
            crcCode[1] = commonSectionContent[crcStartLocation + 1];
            crcCode[2] = commonSectionContent[crcStartLocation + 2];
            crcCode[3] = commonSectionContent[crcStartLocation + 3];
            programAssociationSection.setCrcCode(crcCode);

            mProgramAssociationSection = programAssociationSection;
            mSectionListener.makePat(mProgramAssociationSection);
        }
    }

    @Override
    public void update(Observable observable, Object o) {
        TsPacket tsPacket = (TsPacket) o;
        if (tsPacket.getPid() != PAT_PID) {
            return;
        }
        getSection(tsPacket, SPECIFIED_PAT_TABLE_ID);
    }
}

File: newpaserapp/libtosmartparse/src/main/java/com/example/libtosmartparse/section/ProgramAssociationSectionManager.kt
package com.example.libtosmartparse.section

import com.example.libtosmartparse.packet.TsPacket
import com.example.libtosmartparse.packet.shl
import com.example.libtosmartparse.packet.shr
import java.util.*
import kotlin.experimental.and

/**
 * @author xxm
 * 2022/7/17
 **/
class ProgramAssociationSectionManager : SectionManager(), Observer {
    private var mProgramAssociationSection: ProgramAssociationSection? = null

    override fun update(observable: Observable?, o: Any) {
        val tsPacket: TsPacket = o as TsPacket
        if (tsPacket.getPid() != PAT_PID) {
            return
        }
        getSection(tsPacket, SPECIFIED_PAT_TABLE_ID)
    }

    companion object {
        private const val PROGRAM_ASSOCIATION_SPECIAL_HEAD_LENGTH = 5
        private const val CRC_CODE_LENGTH = 4
        private const val SPECIFIED_PAT_TABLE_ID = 0
        private const val PAT_PID = 0
    }

    override fun composeSpecifiedSection(commonSection: Section?, pid: Int?) {
        if (commonSection!!.getCurrentLength() == commonSection.getSectionLength()) {
            val commonSectionContent = commonSection.getData()
            val programAssociationFilledHead = IntArray(8)
            programAssociationFilledHead[0] = commonSection.getTableId()
            programAssociationFilledHead[1] = commonSection.getReserved()
            programAssociationFilledHead[2] = commonSection.getSectionLength()
            programAssociationFilledHead[3] = commonSectionContent!![0] shl 8 or
                    commonSectionContent[1].toInt() and 0xFF
            programAssociationFilledHead[4] = commonSectionContent[2] shr 1 and 0x1F
            programAssociationFilledHead[5] = (commonSectionContent[2] and 0x01).toInt()
            programAssociationFilledHead[6] = commonSectionContent[3].toInt()
            programAssociationFilledHead[7] = commonSectionContent[4].toInt()
            val programAssociationSection = ProgramAssociationSection()
            programAssociationSection.setHeadDate(programAssociationFilledHead)
            val copyProgramLength =
                commonSectionContent.size - PROGRAM_ASSOCIATION_SPECIAL_HEAD_LENGTH - CRC_CODE_LENGTH
            val programContent = ByteArray(copyProgramLength)
            System.arraycopy(
                commonSectionContent,
                PROGRAM_ASSOCIATION_SPECIAL_HEAD_LENGTH,
                programContent,
                0,
                copyProgramLength
            )
            programAssociationSection.setData(programContent)
            val crcStartLocation = commonSectionContent.size - CRC_CODE_LENGTH
            val crcCode = ByteArray(4)

            crcCode[0] = commonSectionContent[crcStartLocation]
            crcCode[1] = commonSectionContent[crcStartLocation + 1]
            crcCode[2] = commonSectionContent[crcStartLocation + 2]
            crcCode[3] = commonSectionContent[crcStartLocation + 3]
            programAssociationSection.setCrcCode(crcCode)
            mProgramAssociationSection = programAssociationSection
            mSectionListener!!.makePat(mProgramAssociationSection)
        }
    }
}
